File: typescript/src/middleware/hmacMiddleware.ts
import { Request, Response, NextFunction } from 'express';
import { HmacService } from '../services/hmacService';

export const hmacMiddleware = (req: Request, res: Response, next: NextFunction): void => {
  // Skip HMAC validation for GET requests
  if (req.method === 'GET') {
    next();
    return;
  }

  // Get HMAC signature from header
  // Express.js normalizes header names to lowercase
  const hmacSignature = req.headers['x-hmac-signature'] as string;

  // Read request body
  const requestBody = req.body ? JSON.stringify(req.body) : '';

  // Debug: Log the request body and HMAC signature
  console.log('Request Body:', requestBody);
  console.log('HMAC Signature:', hmacSignature);

  // Validate HMAC
  if (!HmacService.validateHmac(requestBody, hmacSignature)) {
    console.log('HMAC validation failed');
    res.status(401).json({
      success: false,
      errorCode: 'INVALID_HMAC_SIGNATURE',
      message: 'Invalid HMAC signature'
    });
    return;
  }

  console.log('HMAC validation passed');

  next();
}; 

File: csharp/SampleGameBackend/Middleware/HmacMiddleware.cs
using Microsoft.AspNetCore.Http;
using SampleGameBackend.Services;
using System.Text;

namespace SampleGameBackend.Middleware
{
    public class HmacMiddleware
    {
        private readonly RequestDelegate _next;

        public HmacMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task InvokeAsync(HttpContext context)
        {
            // Skip HMAC validation for GET requests
            if (context.Request.Method == "GET")
            {
                await _next(context);
                return;
            }

            // Read request body
            context.Request.EnableBuffering();
            var body = await new StreamReader(context.Request.Body).ReadToEndAsync();
            context.Request.Body.Position = 0;

            // Get HMAC signature from header
            var hmacSignature = context.Request.Headers["X-HMAC-Signature"].FirstOrDefault();

            // Validate HMAC
            if (!HmacService.ValidateHmac(body, hmacSignature))
            {
                context.Response.StatusCode = 401;
                await context.Response.WriteAsJsonAsync(new
                {
                    success = false,
                    errorCode = "INVALID_HMAC_SIGNATURE",
                    message = "Invalid HMAC signature"
                });
                return;
            }

            await _next(context);
        }
    }
} 